### PR TITLE
Remove use of non-existent EvalDir option

### PR DESCRIPTION
**Pull request by ghost**
Allow all options to be converted to string. Without this, restoring of the options in **multi_think.cpp** can't work.

https://github.com/yaneurao/YaneuraOu/blob/c221e49425e0d9862b17417f09ca0edb5fa6db10/source/learn/multi_think.cpp#L119-L120

### CODE REVIEW

**Comment by Lancali** [comment]
1. What is the effect of multithreading on data generation? 

```
setoption name Threads value 127
setoption name Hash value 65536
ucinewgame
isready
gensfen depth 12 loop 1000000000 output_file_name generated_kifu.bin
```

2. What is the effect of multithreading on training? 

```
uci
setoption name Threads value 64
setoption name SkipLoadingEval value true 
ucinewgame
isready
learn targetdir training_data loop 100 batchsize 1000000 lambda 0.5 eta 01.0 newbob_decay 0.5 eval_save_interval 100000000 loss_output_interval 1000000 mirror_percentage 50 validation_set_file_name generated_kifu.bin nn_batch_size 1000 eval_limit 3000
```

**Comment by yaneurao** [comment]
2.
There are virtually no negative effects; it simply makes the training faster.

1.
When using multithreading, if the transposition table is shared, the search quality degrades. If you generate data using multiple processes (one process with one thread × N in parallel), the data is generated as intended. However, there is some overhead due to multiprocessing.

**Comment by Lancali** [comment]
> This is correct for nnue-pytorch.
> In the official YaneuraOu trainer, however, it calls qsearch() while training the evaluation parameters.

I asked this question in the nodchip repository, but the answer is different from what you said.

> _In **training**, the network predicts from features and the **loss is computed vs those stored targets**. The classic eval isn’t needed during training/validation unless you explicitly enable a **hybrid** scheme. Mixing versions (generate with SF10/SF14, train with SF11) is fine so long as **feature encoding and conventions match** (HalfKP variant, flip/rotate, score scaling, mate handling, etc.)._


 _Originally posted by @RJurjevic in [#333](https://github.com/nodchip/Stockfish/issues/333#issuecomment-3508194472)_

**Comment by yaneurao** [comment]
As you said, the engine’s normal search is N-thread × 1.
However, only during teacher-data generation do we run 1-thread × N in parallel to improve efficiency.

In YaneuraOu’s source code, this corresponds to Learner::search(), which provides the search functionality for a single thread. By calling this function from multiple threads, we achieve the 1-thread × N parallel configuration.

However, I am planning to clean up the entire namespace Learner soon.
These days the proper learning framework is nnue-pytorch, and for generating training positions, I believe we should use the Python-based GenSfen script instead.

**Comment by yaneurao** [comment]
Yes, your understanding is correct.

**Comment by Lancali** [comment]
@yaneurao Thanks for the very well-written explanations. I reviewed the MultiThinkGenSfen function.

```cpp
				// PVの指し手でleaf nodeまで進めて、そのleaf nodeでevaluate()を呼び出した値を用いる。
				auto evaluate_leaf = [&](Position& pos , vector<Move>& pv)
				{
					auto rootColor = pos.side_to_move();

					int ply2 = ply;
					for (auto m : pv)
					{
						// デバッグ用の検証として、途中に非合法手が存在しないことを確認する。
						// NULL_MOVEはこないものとする。

						pos.do_move(m, states[ply2++]);
						
					}

					// leafに到達
					//      cout << pos;

					auto v = Eval::evaluate(pos);
					// evaluate()は手番側の評価値を返すので、
					// root_colorと違う手番なら、vを反転させて返さないといけない。
					if (rootColor != pos.side_to_move())
						v = -v;

					// 巻き戻す。
					// C++x14にもなって、いまだreverseで回すforeachすらないのか…。
					//  for (auto it : boost::adaptors::reverse(pv))

					for (auto it = pv.rbegin(); it != pv.rend(); ++it)
						pos.undo_move(*it);

					return v;
				};
```

May I ask what exactly does the leaf evaluation function do?

**Comment by Lancali** [comment]
> If by “the leaf evaluation function” you mean qsearch(), then yes 

No, I mean exactly the `evaluate_leaf` function, which is used in the **GenSfen** function.

```cpp
psv.score = evaluate_leaf(pos, pv1);
```

**Comment by Sopel97** [comment]
just block that spammer

**Comment by Lancali** [comment]
> If we need to replace the original data during preprocessing, we also have to keep the original teacher data, which means the amount of training data doubles.

if I understood correctly, to remove the dependence on evaluation functions during training, we need original training data in addition to the current training data. May I ask what do you mean about original data? How to create it?

> Since the teacher data is extremely large (even five years ago I was generating it on the scale of 10 billion positions), doubling that size was unacceptable for me.

I want to train network with only 200 million positions, so doubling that size is acceptable for me. My goal is just to prevent the network from using evaluation functions during learning. Is that possible?

_P.S. FYI Sopel, I am just trying to implement NNUE on Stockfish 10. I'm just asking my question about HalfKP trainer._ 
_https://github.com/HoaryFox/Hoaryfox_

**Comment by Lancali** [comment]
> The team Suisho, which won WCSC35 (the world computer shogi championship held this year), trained its evaluation function using nnue-pytorch.

You're right. However, I prefer to use HalfKP because I think chess/shogi is currently not a solved game. 

> There is no way that nnue-pytorch is flawed.

There is no flawless structure. The stronger the structure, the more weaknesses it has.
